runtime-sdk/src/dispatcher.rs

```diff
@@ -467,8 +447,148 @@ impl<R: Runtime + Send + Sync> transaction::dispatcher::Dispatcher for Dispatche
+                    }
+                }
+                if prefetch_enabled {
+                    ctx.runtime_state()
+                        .prefetch_prefixes(prefixes.into_iter().collect(), R::PREFETCH_LIMIT);
+                }
+
+                // Execute the batch.
+                let mut results = Vec::with_capacity(batch.len());
+                for (index, (tx_size, tx)) in txs.into_iter().enumerate() {
+                    results.push(Self::execute_tx(ctx, tx_size, tx, index)?);
+                }
+
+                Ok(results)
+            },
+        )
+    }
+
+    fn schedule_and_execute_batch(
+        &self,
+        rt_ctx: transaction::Context<'_>,
+        batch: &mut TxnBatch,
+    ) -> Result<ExecuteBatchResult, RuntimeError> {
+        let cfg = R::SCHEDULE_CONTROL.unwrap(); // Must succeed otherwise we wouldn't be here.
+        let mut tx_reject_hashes = Vec::new();
+
+        let mut result = self.execute_batch_common(
+            rt_ctx,
+            |ctx| -> Result<Vec<ExecuteTxResult>, RuntimeError> {
+                // Schedule and execute the batch.
+                //
+                // The idea is to keep scheduling transactions as long as we have some space
+                // available in the block as determined by gas use.
+                let mut new_batch = Vec::new();
+                let mut results = Vec::with_capacity(batch.len());
+                let mut requested_batch_len = cfg.initial_batch_size;
+                'batch: loop {
+                    // Remember length of last batch.
+                    let last_batch_len = batch.len();
+                    let last_batch_tx_hash = batch.last().map(|raw_tx| Hash::digest_bytes(raw_tx));
+
+                    for raw_tx in batch.drain(..) {
+                        // If we don't have enough gas for processing even the cheapest transaction
+                        // we are done.
+                        let remaining_gas = R::Core::remaining_batch_gas(ctx);
+                        if remaining_gas < cfg.min_remaining_gas {
+                            break 'batch;
+                        }
+
+                        // Decode transaction.
+                        let tx = match Self::decode_tx(ctx, &raw_tx) {
+                            Ok(tx) => tx,
+                            Err(_) => {
+                                // Transaction is malformed, make sure it gets removed from the
+                                // queue and don't include it in a block.
+                                tx_reject_hashes.push(Hash::digest_bytes(&raw_tx));
+                                continue;
+                            }
+                        };
+                        let tx_size = raw_tx.len().try_into().unwrap();
+
+                        // If we don't have enough gas remaining to process this transaction, just
+                        // skip it.
+                        if tx.auth_info.fee.gas > remaining_gas {
```

Comment:
:information_source: we always compare the declared gas limit. in the flood-of-max-gas-txs-that-end-up-spending-only-a-little case, we should be able to process one per round. this check lets us skip through the rest in this round and process smaller transactions in the queue